perception/autoware_map_based_prediction/src/map_based_prediction_node.cpp

```diff
@@ -2638,11 +2587,105 @@ std::vector<std::pair<PosePath, double>> MapBasedPredictionNode::convertPathType
     // interpolation for xy
     const auto resampled_converted_path = autoware::motion_utils::resamplePoseVector(
       converted_path, reference_path_resolution_, use_akima_spline_for_xy, use_lerp_for_z);
-    converted_paths.push_back(std::make_pair(resampled_converted_path, width));
+    converted_path_and_width = std::make_pair(resampled_converted_path, width);
   }
 
-  lru_cache_of_convert_path_type_.put(paths, converted_paths);
-  return converted_paths;
+  lru_cache_of_convert_path_type_.put(path, converted_path_and_width);
+  return converted_path_and_width;
+}
+
+std::optional<size_t> MapBasedPredictionNode::searchProperStartingRefPathIndex(
+  const TrackedObject & object, const PosePath & pose_path) const
+{
+  std::unique_ptr<ScopedTimeTrack> st1_ptr;
+  if (time_keeper_) st1_ptr = std::make_unique<ScopedTimeTrack>(__func__, *time_keeper_);
+
+  bool is_position_found = false;
+  std::optional<size_t> opt_index{std::nullopt};
+  auto & index = opt_index.emplace();
+
+  // starting segment index is a segment close enough to the object
+  const auto obj_point = object.kinematics.pose_with_covariance.pose.position;
+  {
+    std::unique_ptr<ScopedTimeTrack> st2_ptr;
+    if (time_keeper_)
+      st2_ptr = std::make_unique<ScopedTimeTrack>("find_close_segment_index", *time_keeper_);
+    double min_dist_sq = std::numeric_limits<double>::max();
+    constexpr double acceptable_dist_sq = 1.0;  // [m2]
+    for (size_t i = 0; i < pose_path.size(); i++) {
+      const double dx = pose_path.at(i).position.x - obj_point.x;
+      const double dy = pose_path.at(i).position.y - obj_point.y;
+      const double dist_sq = dx * dx + dy * dy;
+      if (dist_sq < min_dist_sq) {
+        min_dist_sq = dist_sq;
+        index = i;
+      }
+      if (dist_sq < acceptable_dist_sq) {
+        break;
+      }
+    }
+  }
```

Comment:
- finding the closest point on pose_path to obj_point that comes first within dist_sq
- from there, find the best pose upto search_distance [m]